{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d7bd5b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "ee4651a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def resultados_lncRNA(resultFinal, nameI):\n",
    "    lncRNA = resultFinal[resultFinal['queryName']==\"lncRNA\"]\n",
    "    lncRNA = lncRNA.reset_index(drop=True)\n",
    "    if lncRNA.empty == False:\n",
    "        lncRNA.to_csv(nameI+'_lncRNAs.csv', index=False)\n",
    "        gfflncRNA = salvar_gff(lncRNA, nameI+'_lncRNA')\n",
    "        QtdTotallncRNA = lncRNA['ncRNA'].value_counts(dropna=False, sort=True)\n",
    "        QtdTotallncRNA = QtdTotallncRNA.to_frame().reset_index()\n",
    "        QtdTotallncRNA = QtdTotallncRNA.rename(columns= {'ncRNA': 'Qtd', 'index':'ncRNA'})\n",
    "        \n",
    "        lncRNA2 = lncRNA[['ncRNA', 'IDdb']]\n",
    "        lncRNA2 = lncRNA2.drop_duplicates()\n",
    "        lncRNA2 = lncRNA2.set_index('ncRNA')\n",
    "        lncRNAdict = lncRNA2.to_dict()\n",
    "        QtdTotallncRNA['IDdb'] = QtdTotallncRNA['ncRNA'].map(lncRNAdict['IDdb'])\n",
    "        \n",
    "        QtdTotallncRNA.to_csv(nameI+'_Tabela_quantidade_lncRNAs.csv')\n",
    "        if salvar_Fasta != None:\n",
    "            salvar_fasta(gfflncRNA, nameI+'_lncRNA', list(lncRNA['ncRNA']))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "52af3d62",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Salva em GFF        \n",
    "def salvar_gff(resultFinalFinal, nameI):      \n",
    "    if os.path.exists(nameI+'_ResultadoFinal.gff'):\n",
    "        os.remove(nameI+'_ResultadoFinal.gff')\n",
    "    \n",
    "    gff = pd.DataFrame(columns=['seq_id', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes'])\n",
    "    gff['seq_id'] = resultFinalFinal['targetName']\n",
    "    gff['source'] = resultFinalFinal['source']\n",
    "    gff['type'] = resultFinalFinal['queryName']\n",
    "    gff['start'] = resultFinalFinal['seqFrom']\n",
    "    gff['end'] = resultFinalFinal['seqTo']\n",
    "    gff['score'] = resultFinalFinal['score']\n",
    "    gff['strand'] = resultFinalFinal['strand']\n",
    "    gff['phase'] = '.'\n",
    "    gff['attributes'] = resultFinalFinal['ID'] #resultado do blast/infernal\n",
    "    gff['attributes']=gff['attributes'].str.replace('id', 'ID=id')\n",
    "    gff.to_csv(nameI+'_ResultadoFinal.gff', sep='\\t',mode='a', index=False, header=False)\n",
    "    \n",
    "    file = open(nameI+'_ResultadoFinal.gff', 'r')\n",
    "    lines = file.readlines()\n",
    "    file.close()\n",
    "    lines.insert(0, \"##gff-version 3\"+'\\n')\n",
    "    \n",
    "    file = open(nameI+\"_ResultadoFinal.gff\", 'w')\n",
    "    file.writelines(lines)\n",
    "    file.close()\n",
    "    \n",
    "    return gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "b6d3263f",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Salvar sequencia dos ncRNAs em FASTA\n",
    "def salvar_fasta(gff, nameI, miRNA):\n",
    "    ids = list(gff['seq_id'])\n",
    "    start = list(gff['start'])\n",
    "    end = list(gff['end'])\n",
    "    attr = list(gff['attributes'].str.replace('ID=id',''))\n",
    "    attrOri = list(gff['attributes'])\n",
    "    tipo = list(gff['type'])\n",
    "    i=1\n",
    "    frags=[]\n",
    "    for record in SeqIO.parse('Beatriz/NCBI/Bmusculus/GCF_009873245.2_mBalMus1.pri.v3_genomic.fna', \"fasta\"):\n",
    "        for i in range(len(ids)):\n",
    "            if record.id == ids[i]:\n",
    "                frag = record.seq[start[i]-1:end[i]]\n",
    "                gravar = SeqRecord(frag, \"fragment_\" + str(attr[i]), ids[i], tipo[i]+';'+attrOri[i]+';'+miRNA[i])\n",
    "                frags.append(gravar)\n",
    "    SeqIO.write(frags, nameI+\"_ncRNA.fa\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "33117ce1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetName</th>\n",
       "      <th>seqFrom</th>\n",
       "      <th>seqTo</th>\n",
       "      <th>pident</th>\n",
       "      <th>qcover</th>\n",
       "      <th>IDdb</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>Evalue</th>\n",
       "      <th>ncRNA</th>\n",
       "      <th>queryName</th>\n",
       "      <th>source</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq_1</td>\n",
       "      <td>66769</td>\n",
       "      <td>66840</td>\n",
       "      <td>CL00001</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00005</td>\n",
       "      <td>-</td>\n",
       "      <td>36.8</td>\n",
       "      <td>9.000000e-01</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seq_1</td>\n",
       "      <td>94504</td>\n",
       "      <td>94577</td>\n",
       "      <td>CL00001</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00005</td>\n",
       "      <td>-</td>\n",
       "      <td>35.1</td>\n",
       "      <td>2.700000e+00</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Seq_1</td>\n",
       "      <td>97105</td>\n",
       "      <td>97177</td>\n",
       "      <td>CL00001</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00005</td>\n",
       "      <td>+</td>\n",
       "      <td>39.5</td>\n",
       "      <td>1.700000e-01</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Seq_1</td>\n",
       "      <td>103374</td>\n",
       "      <td>103444</td>\n",
       "      <td>CL00001</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00005</td>\n",
       "      <td>+</td>\n",
       "      <td>39.5</td>\n",
       "      <td>1.600000e-01</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seq_1</td>\n",
       "      <td>193161</td>\n",
       "      <td>193233</td>\n",
       "      <td>CL00001</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00005</td>\n",
       "      <td>-</td>\n",
       "      <td>46.4</td>\n",
       "      <td>1.900000e-03</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29499</th>\n",
       "      <td>Seq_9</td>\n",
       "      <td>107362057</td>\n",
       "      <td>107362163</td>\n",
       "      <td>CL00009</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00026</td>\n",
       "      <td>+</td>\n",
       "      <td>77.4</td>\n",
       "      <td>1.200000e-13</td>\n",
       "      <td>U6</td>\n",
       "      <td>splicing</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id29500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29500</th>\n",
       "      <td>Seq_9</td>\n",
       "      <td>107415824</td>\n",
       "      <td>107415896</td>\n",
       "      <td>CL00001</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00005</td>\n",
       "      <td>-</td>\n",
       "      <td>54.1</td>\n",
       "      <td>7.900000e-06</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id29501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29501</th>\n",
       "      <td>Seq_9</td>\n",
       "      <td>107416039</td>\n",
       "      <td>107416113</td>\n",
       "      <td>CL00001</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00005</td>\n",
       "      <td>-</td>\n",
       "      <td>31.9</td>\n",
       "      <td>1.300000e+01</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id29502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29502</th>\n",
       "      <td>Seq_92</td>\n",
       "      <td>47187</td>\n",
       "      <td>47259</td>\n",
       "      <td>CL00001</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00005</td>\n",
       "      <td>+</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.300000e-04</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id29503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29503</th>\n",
       "      <td>Seq_96</td>\n",
       "      <td>40459</td>\n",
       "      <td>40524</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>RF00994</td>\n",
       "      <td>+</td>\n",
       "      <td>70.7</td>\n",
       "      <td>7.100000e-13</td>\n",
       "      <td>mir-1255</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>Infernal</td>\n",
       "      <td>id29504</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29504 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      targetName    seqFrom      seqTo   pident qcover     IDdb strand  score  \\\n",
       "0          Seq_1      66769      66840  CL00001      -  RF00005      -   36.8   \n",
       "1          Seq_1      94504      94577  CL00001      -  RF00005      -   35.1   \n",
       "2          Seq_1      97105      97177  CL00001      -  RF00005      +   39.5   \n",
       "3          Seq_1     103374     103444  CL00001      -  RF00005      +   39.5   \n",
       "4          Seq_1     193161     193233  CL00001      -  RF00005      -   46.4   \n",
       "...          ...        ...        ...      ...    ...      ...    ...    ...   \n",
       "29499      Seq_9  107362057  107362163  CL00009      -  RF00026      +   77.4   \n",
       "29500      Seq_9  107415824  107415896  CL00001      -  RF00005      -   54.1   \n",
       "29501      Seq_9  107416039  107416113  CL00001      -  RF00005      -   31.9   \n",
       "29502     Seq_92      47187      47259  CL00001      -  RF00005      +   37.0   \n",
       "29503     Seq_96      40459      40524        -      -  RF00994      +   70.7   \n",
       "\n",
       "             Evalue     ncRNA queryName    source       ID  \n",
       "0      9.000000e-01      tRNA      tRNA  Infernal      id1  \n",
       "1      2.700000e+00      tRNA      tRNA  Infernal      id2  \n",
       "2      1.700000e-01      tRNA      tRNA  Infernal      id3  \n",
       "3      1.600000e-01      tRNA      tRNA  Infernal      id4  \n",
       "4      1.900000e-03      tRNA      tRNA  Infernal      id5  \n",
       "...             ...       ...       ...       ...      ...  \n",
       "29499  1.200000e-13        U6  splicing  Infernal  id29500  \n",
       "29500  7.900000e-06      tRNA      tRNA  Infernal  id29501  \n",
       "29501  1.300000e+01      tRNA      tRNA  Infernal  id29502  \n",
       "29502  2.300000e-04      tRNA      tRNA  Infernal  id29503  \n",
       "29503  7.100000e-13  mir-1255     miRNA  Infernal  id29504  \n",
       "\n",
       "[29504 rows x 13 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arq = pd.read_csv('Beatriz/NCBI/Bmusculus/BmusculusFinal_Resultado_Final.csv', skiprows=1, names=['targetName', 'seqFrom', 'seqTo', 'pident', 'qcover', 'IDdb', 'strand', 'score', 'Evalue','ncRNA','queryName', 'source', 'ID'])\n",
    "#arq['targetName'] = arq['targetName'].map(dici)\n",
    "arq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "c60a093a",
   "metadata": {},
   "outputs": [],
   "source": [
    "salvar_Fasta=1\n",
    "resultados_lncRNA(arq,'Beatriz/NCBI/Bmusculus/BmusculusFinal')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d77a12e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26f54bbc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b5618b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dcd8974",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "ca8bf0d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def resultados_miRNA(resultFinal, nameI):\n",
    "    miRNA = resultFinal[(resultFinal['queryName']==\"miRNA\") | (resultFinal['queryName']==\"pre-miRNA\")]\n",
    "    miRNA = miRNA.reset_index(drop=True)\n",
    "    if miRNA.empty == False:\n",
    "        miRNA.to_csv(nameI+'_miRNAs.csv', index=False)\n",
    "        gffmiRNA = salvar_gff(miRNA, nameI+'_miRNA')\n",
    "        QtdTotalmiRNA = miRNA['ncRNA'].value_counts(dropna=False, sort=True)\n",
    "        QtdTotalmiRNA = QtdTotalmiRNA.to_frame().reset_index()\n",
    "        QtdTotalmiRNA = QtdTotalmiRNA.rename(columns= {'ncRNA': 'Qtd', 'index':'ncRNA'})\n",
    "        \n",
    "        miRNA2 = miRNA[['ncRNA', 'IDdb']]\n",
    "        miRNA2 = miRNA2.drop_duplicates()\n",
    "        miRNA2 = miRNA2.set_index('ncRNA')\n",
    "        miRNAdict = miRNA2.to_dict()\n",
    "        QtdTotalmiRNA['IDdb'] = QtdTotalmiRNA['ncRNA'].map(miRNAdict['IDdb'])\n",
    "        \n",
    "        QtdTotalmiRNA.to_csv(nameI+'_Tabela_quantidade_miRNAs.csv')\n",
    "        if salvar_Fasta != None:    \n",
    "            salvar_fasta(gffmiRNA, nameI+'_miRNA', list(miRNA['ncRNA']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3e967b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "resultados_miRNA(arq,'Beatriz/NCBI/Bmusculus/BmusculusFinal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "c6b86457",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import pickle\n",
    "\n",
    "argsFasta = 'Beatriz/NCBI/Bmusculus/GCF_009873245.2_mBalMus1.pri.v3_genomic.fna'\n",
    "outargsFasta = 'Beatriz/NCBI/Bmusculus/New_GCF_009873245.2_mBalMus1.pri.v3_genomic.fna'\n",
    "\n",
    "frags=[]\n",
    "ids=[]\n",
    "n=1\n",
    "for record in SeqIO.parse(argsFasta, 'fasta'):\n",
    "    ids.append(record.id)\n",
    "    frag = record.seq\n",
    "    gravar = SeqRecord(frag, 'Seq_'+str(n), \"\", \"\")\n",
    "    frags.append(gravar)\n",
    "    n=n+1\n",
    "SeqIO.write(frags, outargsFasta, 'fasta')\n",
    "\n",
    "seqDmel=[]\n",
    "for i in range(len(ids)):\n",
    "    seqDmel.append('Seq_'+str(i+1))\n",
    "dfDmel = pd.DataFrame(columns=['Seq','Original'])\n",
    "dfDmel.Original=ids\n",
    "dfDmel.Seq = seqDmel\n",
    "\n",
    "dici=dict(dfDmel.values)\n",
    "\n",
    "with open(\"Beatriz/NCBI/Bmusculus/FASTA_Temp.pkl\", \"wb\") as tf:\n",
    "    pickle.dump(dici,tf)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
